Add details to README (#339) (#3)
ZcashFoundation/frost-zcash-demo: README.md

# Frost Trusted Dealer Demo

A CLI demo for running trusted dealer key generation with the Zcash Foundation's Rust implementation of ['Two-Round Threshold Schnorr Signatures with FROST'](https://datatracker.ietf.org/doc/draft-irtf-cfrg-frost/).

This will be part of a set of demos and a proof of concept application that uses the FROST libraries and reference implementation. The purpose of this is:

1. To identify gaps in our documentation and provide usage examples for developer facing documentation.
2. To provide reference implementations for developers wanting to use FROST in a “real world” scenario.

This demo uses the (Ed25519, SHA-512) ciphersuite. The crate can be found here: https://crates.io/crates/frost-ed25519

## Status ⚠

The Trusted Dealer demo is a WIP.

## Usage

NOTE: This is for demo purposes only and should not be used in production.

> On startup, the Trusted Dealer demo will prompt for:

Reviewer: It would be good to have details of how to run the demo exactly. Assume no prior knowledge of rust or anything else :)

Reviewer: e.g to run this demo, you need to 1. clone this repo, 2. build the demo and 3. run the demo.

Including the commands to do so for each step. This may go in a different separate section. 

I would also add some screenshots of the expected output.

Author: Done

1. Minimum number of signers (>= 2) i.e. The threshold number of participants for the secret sharing scheme
2. Maximum number of signers i.e. the number of shares to generate

The dealer CLI will then use that data to generate:

1. The group public key
2. A commitment to the secret
3. Each participant’s public key

> This above data will be output by the Trusted Dealer demo to the terminal.

Reviewer: Maybe we want to make some comment about how in a "real world" scenario, the secret share should be delivered securely to each participant and not simply printed out to terminal. Additionally, or instead, we might want to explain what each of these items which are output is used for in the next round; which should be kept secret and which should be kept private and how they would ideally be delivered to participants.

I know this is all in the spec but people are lazy about reading links :D

Author: Done

> # About FROST (Flexible Round-Optimised Schnorr Threshold signatures)

Reviewer: _Optional_: I keep going back and forth about whether this section should be right at the end or closer to the top.

Author: Yeh I've moved it to the top under the main description. People will scroll down for more info as it's important context before continuing

Unlike signatures in a single-party setting, threshold signatures require cooperation among a threshold number of signers, each holding a share of a common private key. The security of threshold
schemes in general assume that an adversary can corrupt strictly fewer than a threshold number of participants.

['Two-Round Threshold Schnorr Signatures with FROST'](https://datatracker.ietf.org/doc/draft-irtf-cfrg-frost/) presents a variant of a Flexible Round-Optimized Schnorr Threshold (FROST) signature scheme originally defined in [FROST20](https://eprint.iacr.org/2020/852.pdf). FROST reduces network overhead during threshold
signing operations while employing a novel technique to protect against forgery attacks applicable to prior Schnorr-based threshold signature constructions. This variant of FROST requires two rounds to compute a signature, and implements signing efficiency improvements described by [Schnorr21](https://eprint.iacr.org/2021/1375.pdf). Single-round signing with FROST is not implemented here.

> ## Pre-commit checks

Reviewer: I don't understand what this is for :) Why would someone interested in the demo want to execute this command.

Author: I've put this in a "developer information" section as suggested.

1. Run `cargo make all`

> ## Coverage

Reviewer: Same comment here as above. Are these details intended for internal developers?

Author: Yes. The README is also for anyone contributing, whether that be other zfnd engineers or a third party.

Reviewer 2: These could be inside a `# Development Information` section or something similar

2. Maximum number of signers i.e. the number of shares to generate

```
> cargo run
   Finished dev [unoptimized + debuginfo] target(s) in 0.05s
   Running 'target/debug/frost-trusted-dealer-demo
The minimum number of signers: (2 or more)
2
The maximum number of signers: (must be greater than minimum number of signers)
5
```

The dealer CLI will then use that data to generate:

1. The group public key
2. A commitment to the secret
3. Each signer's public key
4. Each signer's secret share
5. An identifier for each signer

In a "real world" scenario, the secret share should be delivered securely to each participant. For the purposes of the Trusted Dealer demo only, the above data will be output to the terminal. 

In the next round, all signers will receive the data but only the individual signers will receive their personal secret share. 

## Using the output

To generate a key package the participant requires:

* The signer's identifier
* The signer's secret share
* The signer's public key
* The public signing key that represents the entire group

The commitment is used to verify the signer's secret share and to generate the group commitment.

> The dealer can use a secure broadcast channel to ensure each participant has a consistent view of this commitment OR each participant must [verify the commitments](https://github.com/ZcashFoundation/frost/blob/4055cb9439df2814800c678c8da1760a0f86dc10/frost-core/src/frost/keys.rs#L297). 

Reviewer 2: ```suggestion
The dealer should use a secure broadcast channel to ensure each participant has a consistent view of this commitment and each participant must [verify the commitments](https://github.com/ZcashFoundation/frost/blob/4055cb9439df2814800c678c8da1760a0f86dc10/frost-core/src/frost/keys.rs#L297). 
```

The VSS verification must always be done, this was recently fixed in the spec.

Author: Ah ok, thanks.

The trusted dealer MUST delete the secret_key (used during calculation) and secret_share upon completion.
Use of this method for key generation requires a mutually authenticated secure channel between the dealer and participants to send secret key shares, wherein the channel provides confidentiality and integrity. Mutually authenticated TLS is one possible deployment option.

> NOTE: A signer is a participant that signs and sends the signature to receive the final group signature. A dealer is a participant but not a signer.

Reviewer 2: The phrasing here is a bit confusing. The dealer can also be a signer.

I think we can remove the note, but here's a suggestion (assuming I got its intention)

```suggestion
NOTE: A signer is a participant that generates a signing share which is aggregated with the signing shares of other signers to obtain the final group signature.
```

Author: Yeh that is confusing. I was trying to use the right terminology for a participant and a signer as described [here](https://frost.zfnd.org/dev/rfcs/0001-messages.html). Perhaps I can just leave that bit out though. I do think the distinction between participant and signer is a bit confusing and I was debating how to use it properly in the description.
